src/Pinch/Server.hs

```diff
@@ -0,0 +1,124 @@
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE FlexibleContexts          #-}
+{-# LANGUAGE OverloadedStrings         #-}
+{-# LANGUAGE ScopedTypeVariables       #-}
+{-# LANGUAGE TypeApplications          #-}
+{-# LANGUAGE TypeFamilies              #-}
+
+module Pinch.Server
+  ( ThriftServer(..)
+  , ParseError (..)
+  , Channel (..)
+  , Context
+  , ContextItem
+  , addToContext
+  , lookupInContext
+
+  , createServer
+  , runConnection
+  ) where
+
+import           Control.Exception        (Exception, SomeException, throwIO,
+                                           try, tryJust)
+import           Data.Dynamic             (Dynamic (..), fromDynamic, toDyn)
+import           Data.Proxy               (Proxy (..))
+import           Data.Typeable            (TypeRep, Typeable, typeOf, typeRep)
+
+import qualified Data.HashMap.Strict      as HM
+import qualified Data.Text                as T
+
+import           Pinch.Internal.Exception
+import           Pinch.Internal.Message
+import           Pinch.Internal.Pinchable
+import           Pinch.Internal.RPC
+import           Pinch.Internal.TType
+
+import qualified Pinch.Transport          as T
+
+-- | A `Thrift` server. Takes the context and the request message as input and produces a reply message.
+newtype ThriftServer = ThriftServer { unThriftServer :: Context -> Message -> IO (Maybe Message) }
```

Comment:
Having the `Maybe` in here is not terribly nice. I had to add it to support Oneway calls as those must not send any reply. I am starting to wonder if it maybe makes sense to index `Message` by `MessageType`. That way I could use a type family and compute the correct return type based on it. On the other hand, it would be a backwards-incompatible change. What do you think?

Rough sketch:
```
data Message :: MessageType -> * where
  ...

type family ServerResponse a :: *
type instance ServerResponse Oneway = ()
type instance ServerResponse Call = Message

 newtype ThriftServer = ThriftServer { unThriftServer :: Context -> Message t -> IO (ServerResponse t) }
```

Reply:
Turns out that indexing `Message` by type leads to a lot of changes and boilerplate code ....I instead now added a small wrapper which has slightly less guarantees on the type-level but is also a lot less complex:

```
-- new datatype
data Request out where
  RCall :: !Message -> Request Message
  ROneway :: !Message -> Request ()

-- old version
newtype ThriftServer = ThriftServer { unThriftServer :: Context -> Message -> IO (Maybe Message) }

-- new version
newtype ThriftServer = ThriftServer { unThriftServer :: forall a . Context -> Request a -> IO a }
```